tpcx_bb/queries/q25/tpcx_bb_query_25_sql.py

```diff
@@ -0,0 +1,124 @@
+    benchmark,
+    tpcxbb_argparser,
+    run_query,
+    train_clustering_model
+)
+from dask import delayed
+
+
+# -------- Q25 -----------
+# -- store_sales and web_sales date
+q25_date = "2002-01-02"
+
+N_CLUSTERS = 8
+CLUSTER_ITERATIONS = 20
+N_ITER = 5
+
+
+def get_clusters(client, ml_input_df):
+    import dask_cudf
+
+    ml_tasks = [
+        delayed(train_clustering_model)(df, N_CLUSTERS, CLUSTER_ITERATIONS, N_ITER)
+        for df in ml_input_df.to_delayed()
+    ]
+    results_dict = client.compute(*ml_tasks, sync=True)
+
+    output = ml_input_df.index.to_frame().reset_index(drop=True)
+
+    labels_final = dask_cudf.from_cudf(
+        results_dict["cid_labels"], npartitions=output.npartitions
+    )
+    output["label"] = labels_final.reset_index()[0]
+
+    # Based on CDH6.1 q25-result formatting
+    results_dict["cid_labels"] = output
+    return results_dict
+
+
+def read_tables(data_dir, bc):
+    bc.create_table("web_sales", data_dir + "web_sales/*.parquet")
+    bc.create_table("store_sales", data_dir + "store_sales/*.parquet")
+    bc.create_table("date_dim", data_dir + "date_dim/*.parquet")
+
+
+def main(data_dir, client, bc, config):
+    benchmark(read_tables, data_dir, bc, dask_profile=config["dask_profile"])
+
+    query = f"""
+        WITH concat_table AS
+        (
+            (
+                SELECT
+                    ss_customer_sk AS cid,
+                    count(distinct ss_ticket_number) AS frequency,
+                    max(ss_sold_date_sk) AS most_recent_date,
+                    CAST( SUM(ss_net_paid) AS DOUBLE) AS amount
+                FROM store_sales ss
+                JOIN date_dim d ON ss.ss_sold_date_sk = d.d_date_sk
+                WHERE CAST(d.d_date AS DATE) > DATE '{q25_date}'
+                AND ss_customer_sk IS NOT NULL
+                GROUP BY ss_customer_sk
+            ) union all
+            (
+                SELECT
+                    ws_bill_customer_sk AS cid,
+                    count(distinct ws_order_number) AS frequency,
+                    max(ws_sold_date_sk)   AS most_recent_date,
+                    CAST( SUM(ws_net_paid) AS DOUBLE) AS amount
+                FROM web_sales ws
+                JOIN date_dim d ON ws.ws_sold_date_sk = d.d_date_sk
+                WHERE CAST(d.d_date AS DATE) > DATE '{q25_date}'
+                AND ws_bill_customer_sk IS NOT NULL
+                GROUP BY ws_bill_customer_sk
+            )
+        )
+        SELECT
+            cid AS cid,
+            CASE WHEN 37621 - max(most_recent_date) < 60 THEN 1.0
+                ELSE 0.0 END AS recency, -- 37621 == 2003-01-02
+            CAST( SUM(frequency) AS BIGINT) AS frequency, --total frequency
+            CAST( SUM(amount) AS DOUBLE)    AS amount --total amount
+        FROM concat_table
+        GROUP BY cid
+        ORDER BY cid
+    """
+    result = bc.sql(query)
+
+    # Prepare df for KMeans clustering
+    cluster_input_ddf = result.set_index('cid')
```

Comment:
Do we need to do a global set index here? This will be *very* expensive at scale across nodes.

Reply:
It doesn't have to be right there, you're right. After calling the `repartition()` function I think it would be better, do you think?

```
...
cluster_input_ddf = cluster_input_ddf.repartition(npartitions=1)
cluster_input_ddf = cluster_input_ddf.set_index('cid')
cluster_input_ddf = cluster_input_ddf.persist()

results_dict = get_clusters(client=client, ml_input_df=cluster_input_ddf)
return results_dict
```

Reply:
This would certainly be better. I guess I'm more curious why we need to do it in the first place

Reply:
For a strange reason using   `cluster_input_ddf = cluster_input_ddf.set_index('cid')`  shows a low Cost delta percent 
```
Cost delta percent: 2.6814614820339198e-06
RAPIDS cost lower/similar: True
Label overlap percent: 99.99996819836082
Correctness Assertion True
```

Without `cluster_input_ddf = cluster_input_ddf.set_index('cid')` (and aditional line code) shows a high Cost delta percent
```
Clustering Query
Cost delta percent: 56.63223594034793
RAPIDS cost lower/similar: False
Error 
Correctness Assertion False
```

Reply:
Thanks. I suspect the difference is due to the Dask implementation assuming the index is meaningful (from the groupby). There's likely low-hanging fruit to resolve it in this BSQL query, but in the interest of getting all queries completed let's move forward as is. I'll file an optimization issue for this (and q26).